credoai/lens/lens.py

```diff
@@ -281,30 +313,30 @@ def _add(
 
     def _cycle_add_through_ds_feat(
         self,
-        evaluator,
-        id,
+        pipeline_step,
         check_sens_feat,
         check_data,
         evaluator_arguments,
         features_to_eval,
     ):
         for feat in features_to_eval:
+            additional_meta = {}
+            if check_sens_feat:
+                additional_meta["sensitive_feature"] = feat
             if check_data:
-                available_datasets = [
-                    n for n, a in vars(self).items() if "data" in n if a
-                ]
-                for dataset in available_datasets:
-                    labels = {"sensitive_feature": feat} if check_sens_feat else {}
-                    labels["dataset"] = dataset
-                    evaluator_arguments["data"] = vars(self)[dataset]
+                for dataset_label, dataset in self.get_datasets().items():
+                    additional_meta["dataset"] = dataset_label
+                    step = deepcopy(pipeline_step)
```

Comment:
One thing that occurs to me now that I'm a little bit deeper on the deepchecks evaluator is that this strategy of copying evaluators per-dataset is not always going to give us what we want.

If I'm not mistaken, all the evaluators we have right now either require 'data' (i.e. it's a single-dataset assessment) or BOTH 'assessment' and 'train' data to be specified. In the latter case, things are simple: if you don't provide both, the validation check fails. In the former, however, right now we just run the evaluator on one or both data sets. This line achieves that by copying the pipeline step and specifying that we should use the 2nd dataset if one is present.

Integrating deepchecks into Lens throws a bit of a wrench in this. Some checks in the package only require a single dataset while others require both train and test. Following the existing paradigm of specifying a `required_artifacts` dictionary, we could require either `data` (one data set) or both `assessment_data` and `training_data`.

If we choose the latter, we won't be able to run any of deepchecks's single-dataset checks.
OTOH, if we choose the former, then this snippet of code (lens.py L327-333), as presently written, will make it such that the deepchecks evaluator only ever sees ONE data set at a time and so we wouldn't be able to run any of deepchecks' multi-dataset checks.

Reply:
Update: Workaround 3 (which I'm going to pursue for now): implement a new `check_requirements` function that only raises a `ValidationError` if neither dataset is present